staging/src/k8s.io/apimachinery/pkg/api/meta/help.go

```diff
@@ -19,21 +19,55 @@ package meta
 import (
 	"fmt"
 	"reflect"
+	"sync"
 
 	"k8s.io/apimachinery/pkg/conversion"
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
-// IsListType returns true if the provided Object has a slice called Items
+var (
+	// isListCache maintains a cache of types that are checked for lists
+	// which is used by IsListType.
+	isListCache = struct {
+		lock   sync.RWMutex
+		byType map[reflect.Type]bool
+	}{
+		byType: make(map[reflect.Type]bool, 1024),
+	}
+)
+
+// IsListType returns true if the provided Object has a slice called Items.
 func IsListType(obj runtime.Object) bool {
-	// if we're a runtime.Unstructured, check whether this is a list.
-	// TODO: refactor GetItemsPtr to use an interface that returns []runtime.Object
-	if unstructured, ok := obj.(runtime.Unstructured); ok {
-		return unstructured.IsList()
+	switch t := obj.(type) {
+	case runtime.Unstructured:
+		return t.IsList()
+	}
+	t := reflect.TypeOf(obj)
+
+	isListCache.lock.RLock()
+	ok, exists := isListCache.byType[t]
+	isListCache.lock.RUnlock()
+
+	if !exists {
+		_, err := GetItemsPtr(obj)
+		ok = err == nil
+
+		// we keep the cache at a bounded size out of an abundance of caution
+		isListCache.lock.Lock()
+		if excess := len(isListCache.byType) - 1024; excess > 100 {
```

Comment:
nit: 1024 seems low. We have 238 types with ObjectMeta and 177 with ListMeta, so we're almost halfway to the limit already. 4096 seems safer, especially if eviction is random.

Reply:
I can increase, I don't even think there's an issue with 16k other than possibly lookup time and the fact this is imported by every client in the entire universe, but only used by the apiserver.  Alternative, keep an apiserver specific cache.

Reply:
So I would vote for increasing (and it would be useful to expose a prometheus metric to be able tro track the current size of the cache).

Re importing by client - as long as you don't use it, I don't see a problem with it. And once you start using it, you will also probably benefit from it. So I don't think it's a significant problem.

Reply:
After discussion stayed at 1024 but just fell back to caching only the first 1024, added comments about the real solution here forcing an interface check which was our overall direction anyway.  Reflection must die.